aircmd/models/base.py

```diff
@@ -45,50 +44,143 @@ class GlobalSettings(BaseSettings):
+    parent_pipeline: Optional['Pipeline']
+    dagger_client: dagger.Client
+
+    class Config:
+        arbitrary_types_allowed=True
+
     @classmethod
-    def load_secrets_from_file(cls, secrets_file: str) -> None:
-        """Decrypt the secrets file using SOPS and load it into environment variables."""
-        try:
-            decrypted_secrets = subprocess.check_output(["sops", "-d", secrets_file])
-            with open(".env", "wb") as f:
-                f.write(decrypted_secrets)
-            load_dotenv(".env")
-            os.remove(".env")
-        except subprocess.CalledProcessError as e:
-            raise ValueError(f"Error decrypting secrets file: {e}")
-        except FileNotFoundError:
-            raise ValueError("SOPS command not found. Please install SOPS to use this feature.")
-
-# Mutable. Store pipeline state here. Created at runtime
+    async def create(cls, name: str, parent_pipeline: Optional['Pipeline'] = None) -> 'Pipeline':
+        if parent_pipeline is None:
+            # This is a top-level pipeline. Get the Dagger client from the current PipelineContext.
+            async with PipelineContext() as ctx:
+                dagger_client = ctx.dagger_client
+        else:
+            # This is a nested pipeline. Get the Dagger client from the parent pipeline.
+            dagger_client = parent_pipeline.dagger_client
+
+        # Create the pipeline using the appropriate Dagger client.
+       
+        dagger_client = dagger_client.pipeline(name)
+        return cls(name=name, parent_pipeline=parent_pipeline, dagger_client=dagger_client)
+
+
 class PipelineContext(BaseModel, Singleton):
-    dagger_config: dagger.Config
-    dagger_connection: dagger.Connection
+    _current_level: int = 0
+    logger: logging.Logger
+    current_running_tasks: int = 0
+    max_concurrency: int = 0
+    max_concurrency_per_level: dict[int, int] = {}
+    concurrency_lock: anyio.Lock = anyio.Lock()
+    _dagger_client: Optional[dagger.Client] = PrivateAttr()
+    dagger_client: dagger.Client = Field(default=None, init=False)
 
-    def __init__(self, config: Optional[dagger.Config] = None,  
-                 connection: Optional[dagger.Connection] = None, 
-                 **data: Any):
-        if config is None:
-            config = dagger.Config(log_output=sys.stdout)
-        if connection is None:
-            # Replace with your logic to create a Connection instance
-            connection = dagger.Connection(config = config)
-        super().__init__(dagger_config=config, dagger_connection=connection, **data)
 
     class Config:
-        arbitrary_types_allowed = True
+        arbitrary_types_allowed=True
+
+    
+    async def __aenter__(self) -> 'PipelineContext':
+        self.dagger_client = await dagger.Connection(dagger.Config(log_output=sys.stdout)).__aenter__()
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
+        if self.dagger_client is not None:
```

Comment:
Is this stubbed? Seems like you can pass unconditionally

Reply:
Helder is going to suggest how we properly tear this down, it was indeed a stub and is not being torn down properly